### PR TITLE
Implem and integration of the Serpent cipher (ECB)

### PR DESCRIPTION
**Pull request by NicsTr**
Implementation and integration in rahash2 of the Serpent block cipher (https://www.cl.cam.ac.uk/~rja14/serpent.html). I tested it with some test vectors that can be found here: https://www.cs.technion.ac.il/~biham/Reports/Serpent/.

For now, only the ECB mode of encryption is available.

The non-regression tests PR can be found here: https://github.com/radare/radare2-regressions/pull/1077.

### CODE REVIEW

**Comment by wargio** [comment]
i won't place sboxes into the header.
i would prefer to have them in the source file.

**Comment by NicsTr** [comment]
@wargio 
I'm failing to understand why it would be better.
For your information, I put it in the header because the AES implementation in r2 was done this way.

**Comment by wargio** [comment]
i know that they are in the header, but i prefer having static sboxes (which are only used internally) to be in the source file.
mine was a comment on that, i'll wait pancake to comment this.
Just to be clear: it is not a mistake. you correcly placed them in the header as they should be from previous coding rules.

File: libr/crypto/p/crypto_serpent_algo.c [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	rotl(&tmp_block[0], 5);
+	rotl(&tmp_block[2], 22);
+	apply_IP(tmp_block, block);
+}
+
+void apply_permut_inv(ut32 block[DW_BY_BLOCK]) {
+	ut32 tmp_block[DW_BY_BLOCK] = {0};
+	apply_FP(block, tmp_block);
+	rotr(&tmp_block[0], 5);
+	rotr(&tmp_block[2], 22);
```
Comment (radare) [inline libr/crypto/p/crypto_serpent_algo.c]: Spaces before (

File: libr/crypto/p/crypto_serpent_algo.c [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	tmp_block[2] ^= tmp_block[3]^(tmp_block[1]<<7);
```
Comment (radare) [inline libr/crypto/p/crypto_serpent_algo.c]: Spaces around ^

File: libr/crypto/p/crypto_serpent_algo.c [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+	tmp_block[3] ^= tmp_block[2]^(tmp_block[0]<<3);
+	tmp_block[1] ^= tmp_block[0]^tmp_block[2];
+	rotr(&tmp_block[2], 3); 
+	rotr(&tmp_block[0], 13);
+	apply_IP(tmp_block, block);
+}
+
+void apply_round(int round, ut32 block[DW_BY_BLOCK], 
+		ut32 subkeys[DW_BY_BLOCK*NB_SUBKEYS]) {
+
```
Comment (radare) [inline libr/crypto/p/crypto_serpent_algo.c]: Too many empty lines

File: libr/crypto/p/crypto_serpent_algo.c [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+	
+	ut32 res;
+	int i, j;
+
+	apply_xor(block, &(subkeys[4*round]));
```
Comment (radare) [inline libr/crypto/p/crypto_serpent_algo.c]: Better use subkeys + 4*round, shorter and easier to read

File: libr/crypto/p/crypto_serpent_algo.c [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+		apply_xor(block, &(subkeys[4*(round+1)]));
+	} else {
+		apply_permut_inv(block);
+	}
+
+	int i, j;
+
+	ut32 res;
+	for (i = 0; i < DW_BY_BLOCK; i++) {
+		res = 0;
```
Comment (radare) [inline libr/crypto/p/crypto_serpent_algo.c]: Ut32 here. Assign and define in the same line

File: libr/crypto/p/crypto_serpent_algo.h [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+/*
+ * st: A serpent_state structure containing the key and the key size.
+ * subkeys: When the function returns, an array of double words containings
+ *          all the subkeys needed for the encryptio/dcryption with serpent.
+ */
+void serpent_keyschedule(struct serpent_state st,
+        ut32 subkeys[NB_SUBKEYS*DW_BY_BLOCK]);
+
+static const ut8 S[][16] = {
```
Comment (radare) [inline libr/crypto/p/crypto_serpent_algo.h]: Move this into the C

File: libr/crypto/p/crypto_serpent.c [hunk 7]
```diff
@@ -0,0 +1,107 @@
+#include <r_lib.h>
+#include <r_crypto.h>
+#include "crypto_serpent_algo.h"
+
+static struct serpent_state st = {{0}};
+
+static bool serpent_set_key (RCrypto *cry, const ut8 *key, int keylen, int mode, int direction) {
+	if (!(keylen == 128 / 8 || keylen == 192 / 8 || keylen == 256 / 8)) {
```
Comment (radare) [inline libr/crypto/p/crypto_serpent.c]: Add () on each item

File: libr/crypto/p/crypto_serpent.c [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+		for (i = 0; i < blocks; i++) {
+			// delta in number of ut32
+			const int delta = (BLOCK_SIZE * i)/4;
+			serpent_decrypt (&st, ibuf + delta, tmp + delta);
+		}
+	}
+	
+	// Construct ut32 blocks from byte stream
+	for (j = 0; j < size/4; j++) {
+		obuf[4*j] = tmp[j] & 0xff;
```
Comment (radare) [inline libr/crypto/p/crypto_serpent.c]: Save int k =  j * 4; and reuse that computation instead of obuf... every line

**Comment by radare** [comment]
Looks good to merge after fixing my comments. Thanks!

**Pull request merged**